src/components/ChallengeModal/ChallengeModal.tsx

```diff
@@ -752,10 +742,27 @@ export class ChallengeModalBody extends React.Component<
     };
 
     /* update bindings  */
-    update_conf_bot_id = (ev: React.ChangeEvent<HTMLSelectElement>) =>
-        this.upstate("conf.bot_id", parseInt(ev.target.value));
-    update_challenge_game_name = (ev: React.ChangeEvent<HTMLInputElement>) =>
-        this.upstate(this.gameStateName("name"), ev);
+
+    /* nested fn updates */
+    update_conf = (update_fn: UpdateFn<ChallengeModalConf>): void =>
+        this.setState((prev) => ({ ...prev, conf: update_fn(prev.conf) }));
+    update_challenge_settings = (update_fn: UpdateFn<ChallengeModalChallengeSettings>): void =>
+        this.setState((prev) => ({ ...prev, challenge: update_fn(prev.challenge) }));
+    update_demo_settings = (update_fn: UpdateFn<ChallengeModalDemoSettings>): void =>
+        this.setState((prev) => ({ ...prev, demo: update_fn(prev.demo) }));
+    update_game_settings = (update_fn: UpdateFn<ChallengeModalGameSettings>): void =>
+        this.update_challenge_settings((prev) => ({ ...prev, game: update_fn(prev.game) }));
+
+    /* direct fn updates */
+    update_bot_id = (id: number) => this.update_conf((prev) => ({ ...prev, bot_id: id }));
```

Comment:
Doesn't have to be in this PR, but part of the pain, I'm sure you noticed, is the heavy nesting of `state`.  It may be worthwhile to flatten the state object in the future to make this easier.

https://react.dev/learn/choosing-the-state-structure#avoid-deeply-nested-state

Reply:
Also maybe the nesting can be reduced by extracting (or splitting) parts as individual components. I feel like the readability suffers a lot from the many different use-cases, and the giant file size.